rclcpp/include/rclcpp/context.hpp

```diff
@@ -196,29 +203,38 @@ class Context : public std::enable_shared_from_this<Context>
    * and persist on repeated init's.
    *
    * \param[in] callback the on shutdown callback to be registered
-   * \return the callback passed, for convenience when storing a passed lambda
+   * \return the created shared pointer of callback handler
    */
   RCLCPP_PUBLIC
   virtual
-  OnShutdownCallback
-  on_shutdown(OnShutdownCallback callback);
+  OnShutdownCallbackHandle::SharedPtr
+  on_shutdown(OnShutdownCallbackHandle::OnShutdownCallbackType callback);
```

Comment:
The caller now needs to retain the return value to ensure that the callback was still valid, which wasn't the case before.
It's better to return an identifier that can be passed to remove but that you don't need to retain if you want to keep the callback forever, so previous code isn't broken.

Reply:
i.e. I prefer either the idea of returning an iterator to the handle (which breaks ABI if you change the container type), or what @wjwwood mentioned about using a `std::unordered_map` with an identifier using as a key (incremented each time a new callback is added).

Reply:
> The caller now needs to retain the return value to ensure that the callback was still valid, which wasn't the case before.

Yeah. It has a problem. (Refer to my comments https://github.com/ros2/rclcpp/pull/1639#issuecomment-824654837)